docs/build-provenance-for-all-registries.md

```diff
@@ -0,0 +1,102 @@
+
+As of April 2023, there is a [public beta of build provenance in npm](https://github.blog/2023-04-19-introducing-npm-package-provenance/). We’ll cover how npm implemented this functionality, and use that to frame how to add provenance to other registries.
+
+
+# npm build provenance
+
+Here’s a high-level diagram of how npm build provenance works:
+
+![Diagram of provenance information flows between the cloud CI/CD system, Sigstore, and the npm registry](./imgs/provenance-flow.png)
+
+Let’s break this down into a couple of parts:
+
+- Collecting information from the cloud CI/CD system
+- Extending the registry to support build provenance
+- Supplying build provenance to users for verification
+
+
+# Collecting information from the cloud CI/CD provider
+
+You will need something reusable that the build process can call to get the information from the CI/CD provider, assemble it into a standard format, and then send that information to Sigstore and your registry.
+
+We recommend building this functionality into a CLI tool. Most ecosystems already have some CLI tool involved in the package publishing process, and this allows the implementation to be shared across CI/CD providers.
+
+For detailed technical specifications on assembling provenance and interacting with Sigstore services, see the [Sigstore Client Specification](https://docs.google.com/document/d/1kbhK2qyPPk8SLavHzYSDM8-Ueul9_oxIMVFuWMWKz0E/edit?usp=sharing). Here we’ll go over a high-level description of the steps your tooling will follow (and take a look at [npm CLI’s publish.js](https://github.com/npm/cli/blob/latest/workspaces/libnpmpublish/lib/publish.js) for a reference implementation):
+
+If your registry supports private packages, you should ensure that you’re working on a public package. This is because the build metadata will be published to Sigstore’s transparency log, Rekor, which publishers of private packages might not want to make public.
+
+Then we want to request an OIDC token that Sigstore Fulcio will accept. You should ensure you’re on [CI/CD provider Fulcio supports](https://github.com/sigstore/fulcio/tree/main/federation), and if so, request an OIDC token with the audience set to "sigstore".
+
+Next you’ll construct the build provenance attestations. We recommend using [in-toto attestations](https://github.com/in-toto/attestation/blob/main/spec/README.md#in-toto-attestation-framework-spec) for formatting your attestation document, and [SLSA provenance](https://slsa.dev/provenance/v1) for the predicates in the in-toto attestation document. See [npm CLI's provenance.js](https://github.com/npm/cli/blob/latest/workspaces/libnpmpublish/lib/provenance.js) for a reference implementation.
```

Comment:
I want to note that in-toto/attestation also provides models in [Python](https://github.com/in-toto/attestation/tree/main/python/in_toto_attestation/predicates/provenance/v1), [Go](https://github.com/in-toto/attestation/tree/main/go/predicates/provenance/v1), and [Java](https://github.com/in-toto/attestation/tree/main/java/io/github/intoto/attestation/predicates/provenance/v1) using the official [SLSA Provenance proto](https://github.com/in-toto/attestation/blob/main/protos/in_toto_attestation/predicates/provenance/v1/provenance.proto), with published libraries in the works in those languages.

Reply:
I've added a link to the official SLSA provenance proto.